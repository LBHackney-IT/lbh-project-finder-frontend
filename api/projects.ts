import axios from "axios";
import { Project, ProjectSearchResults } from "../types";

const ENDPOINT_API = process.env.ENDPOINT_API;
const API_KEY = process.env.API_KEY;

const keyHeader = {
  'x-api-key': API_KEY
}

export const getProjects = async (
  params: any
): Promise<ProjectSearchResults> => {
  const { data } = await axios.get(`${ENDPOINT_API}/projects`, {
    headers: keyHeader,
    params: params,
  });

  return data;
};

export const getProject = async (projectId: number): Promise<Project> => {
  const { data } = await axios.get(`${ENDPOINT_API}/projects/${projectId}`, { headers: keyHeader });

  return data;
};

export const addProject = async (formData: Record<string, string>): Promise<any> => {
  const { data } = await axios.post(`${ENDPOINT_API}/projects`, formData, { headers: keyHeader })

  return data;
};

export const updateProject = async (formData: Record<string, string | number>): Promise<void> => {
  await axios.patch(`${ENDPOINT_API}/projects`, formData, { headers: keyHeader });
};

export const deleteProject = async (project_id: number): Promise<void> => {
  await axios.delete(`${ENDPOINT_API}/projects/${project_id}`, { headers: keyHeader });
}
