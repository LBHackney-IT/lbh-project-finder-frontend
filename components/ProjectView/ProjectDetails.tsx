import { Project } from "../../types";

interface Props {
  project?: Project;
}
const ProjectDetails = ({ project }: Props): React.ReactElement => {
  return (
    <div>
      <dl className="govuk-summary-list lbh-summary-list">
        <div className="govuk-summary-list__row">
          <dt className="govuk-summary-list__key">Project Contact</dt>
          <dd className="govuk-summary-list__value">
            {project?.projectContact}
          </dd>
        </div>
        <div className="govuk-summary-list__row">
          <dt className="govuk-summary-list__key">Phase</dt>
          <dd className="govuk-summary-list__value">{project?.phase}</dd>
        </div>
        <div className="govuk-summary-list__row">
          <dt className="govuk-summary-list__key">Size</dt>
          <dd className="govuk-summary-list__value">{project?.size}</dd>
        </div>
        <div className="govuk-summary-list__row">
          <dt className="govuk-summary-list__key">Category</dt>
          <dd className="govuk-summary-list__value">
            {project?.category ? project.category : "N/A"}
          </dd>
        </div>
        <div className="govuk-summary-list__row">
          <dt className="govuk-summary-list__key">Priority</dt>
          <dd className="govuk-summary-list__value">
            {project?.priority ? project.priority : "N/A"}
          </dd>
        </div>
        <div className="govuk-summary-list__row">
          <dt className="govuk-summary-list__key">Product Users</dt>
          <dd className="govuk-summary-list__value">
            {project?.productUsers ? project.productUsers : "N/A"}
          </dd>
        </div>
        <div className="govuk-summary-list__row">
          <dt className="govuk-summary-list__key">Project Dependencies</dt>
          <dd className="govuk-summary-list__value">
            {project?.dependencies ? project?.dependencies : "N/A"}
          </dd>
        </div>
      </dl>
    </div>
  );
};

export default ProjectDetails;
