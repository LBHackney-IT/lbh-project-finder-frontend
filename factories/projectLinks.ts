import { Factory } from "fishery";
import { ProjectLink } from "../types";

export const projectLinkFactory = Factory.define<ProjectLink>(
  ({ sequence }) => ({
    id: sequence,
    project_id: sequence + 1,
    type: "Google Drive",
    link: "https://fakelink.com",
  })
);

export const mockedProjectLink = projectLinkFactory.build();
export const mockedProjectsLinks = [projectLinkFactory.build()];
