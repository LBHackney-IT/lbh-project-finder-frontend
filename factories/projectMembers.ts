import { Factory } from "fishery";
import { ProjectMember } from "../types";

export const projectMemberFactory = Factory.define<ProjectMember>(
  ({ sequence }) => ({
    id: sequence,
    member_id: sequence + 1,
    project_id: sequence + 2,
    project_member: "Some Person",
    role: "Developer",
  })
);

export const mockedProjectMember = projectMemberFactory.build();
export const mockedProjectsMembers = [projectMemberFactory.build()];
