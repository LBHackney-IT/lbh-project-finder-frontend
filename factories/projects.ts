import { Factory } from 'fishery';
import { Project } from '../types';

export const projectFactory = Factory.define<Project>(({ sequence }) => ({
    project_id: sequence,
    project_name: "Test Project",
    description: "A project with many details",
    stage: "Discovery",
    size: "Large",
    type: "Tech",
    project_dependencies: "none",
    project_links: [{
        name: "Foo",
        link: "fake link"
    }
    ],
    project_team: [{ name: "Julius Caesar", role: "Delivery Manager" }]
}
));

export const mockedProject = projectFactory.build();
export const mockedProjects = [projectFactory.build()];