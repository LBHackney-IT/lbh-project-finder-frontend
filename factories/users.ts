import { Factory } from "fishery";

import { User } from "../types";

export const userFactory = Factory.define<User>(() => ({
  name: "guido",
  hasAdminPermissions: false,
  hasUserPermissions: true,
  email: "guido@bar.com",
  isAuthorised: true,
}));

export const mockedAdminUser = userFactory.build({
  hasAdminPermissions: true,
  hasUserPermissions: false,
});

export const mockedUser = userFactory.build();
