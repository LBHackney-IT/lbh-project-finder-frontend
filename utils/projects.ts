import axios, { AxiosError } from "axios";
import useSWR, { SWRResponse } from "swr";
import { Project, ProjectSearchResults, SearchFormData } from "../types";

export const useProjects = async (
  params: SearchFormData
): Promise<ProjectSearchResults> => {
  const { data } = await axios.get(`/api/projects`, { params });
  return data;
};

export const useProject = (
  projectId: number
): SWRResponse<Project, AxiosError> => useSWR(`/api/projects/${projectId}`);

export const addProject = async (
  formData: Record<string, string>
): Promise<Record<string, string>> => {
  const { data } = await axios.post(`/api/projects`, formData);
  return data;
};

export const updateProject = async (
  projectId: number,
  formData: Record<string, string | number>
): Promise<Record<string, string>> => {
  const { data } = await axios.patch(`/api/projects/${projectId}`, formData);
  return data;
};
