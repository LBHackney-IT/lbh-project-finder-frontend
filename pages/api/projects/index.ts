import type { NextApiRequest, NextApiResponse, NextApiHandler } from "next";

import { StatusCodes } from "http-status-codes";
import { addProject, getProjects } from "../../../api/projects";
import { SearchFormData } from "../../../types";

const endpoint: NextApiHandler = async (
  req: NextApiRequest,
  res: NextApiResponse
) => {
  switch (req.method) {
    case "GET":
      try {
        console.log("Query is: " + JSON.stringify(req.query));
        const data = await getProjects(req.query);
        res.status(StatusCodes.OK).json(data);
      } catch (error) {
        console.error("Project get error:", error?.response?.data);
        console.error("Project get request:", req);
        res
          .status(StatusCodes.INTERNAL_SERVER_ERROR)
          .json({ message: "Unable to get project" });
      }
      break;

    case "POST":
      try {
        const data = await addProject(req.body);
        res.status(StatusCodes.OK).json(data);
      } catch (error) {
        console.error("Project post error:", error?.response?.data);
        console.error("Project post request:", req);
        res
          .status(StatusCodes.INTERNAL_SERVER_ERROR)
          .json({ message: "Unable to add project" });
      }
      break;

    default:
      res
        .status(StatusCodes.BAD_REQUEST)
        .json({ message: "Invalid request method" });
  }
};

export default endpoint;
