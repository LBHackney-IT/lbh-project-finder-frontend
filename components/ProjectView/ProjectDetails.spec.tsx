import { render } from '@testing-library/react';
import { mockedProject } from '../../factories/projects';

import ProjectDetails from './ProjectDetails'

describe('ProjectDetails component', () => {
    const props = mockedProject

    it('should render a ProjectDetails component with the correct data', () => {
        const { asFragment } = render(<ProjectDetails project={props} />);

        expect(asFragment()).toMatchSnapshot();
    });
});