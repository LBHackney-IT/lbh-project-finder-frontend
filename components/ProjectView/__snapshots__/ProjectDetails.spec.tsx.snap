// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectDetails component should render a ProjectDetails component with the correct data 1`] = `
<DocumentFragment>
  <div>
    <dl
      class="govuk-summary-list lbh-summary-list"
    >
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          Project Contact
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          Someone
        </dd>
      </div>
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          Phase
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          Discovery
        </dd>
      </div>
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          Size
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          Large
        </dd>
      </div>
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          Category
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          Tech
        </dd>
      </div>
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          Priority
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          Extreme
        </dd>
      </div>
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          Product Users
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          Noone
        </dd>
      </div>
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          Project Dependencies
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          none
        </dd>
      </div>
    </dl>
  </div>
</DocumentFragment>
`;
